course_discovery/apps/core/tests/test_install_es_indexes.py

```diff
@@ -10,25 +10,14 @@
 class InstallEsIndexesCommandTests(ElasticsearchTestMixin, TestCase):
     def test_create_index(self):
         """ Verify the app sets the alias and creates a new index. """
-        index = settings.HAYSTACK_CONNECTIONS['default']['INDEX_NAME']
+        index_names = settings.ELASTICSEARCH_INDEX_NAMES
+        for __, index_name in index_names.items():
+            # Delete the index
+            self.es.indices.delete(index=index_name, ignore=404)
+            self.assertFalse(self.es.indices.exists(index=index_name))
 
-        # Delete the index
-        self.es.indices.delete(index=index, ignore=404)
-        self.assertFalse(self.es.indices.exists(index=index))
-
-        call_command('install_es_indexes')
+        call_command('search_index', '--create')
 
         # Verify the index was created
-        self.assertTrue(self.es.indices.exists(index=index))
-
-    def test_alias_exists(self):
```

Comment:
I deleted this test because I was sure that aliases would no longer be used and they were not needed. Now I doubt that I am 100% right and that I did the right thing.
I here very much ask for your help and knowledge to help me understand 100% how aliases work in this service and why aliases were made.
The main problem is that now the creation of indexes is done in a different way, much easier and more efficient, and creating aliases as it was before is not so easy. We need to think about how to do this in new conditions so that they fulfill their old features. This is now a blocker for me and a real headache. One of the few things left to do.
The ideal scenario would be for us to abandon this functionality at all. It seems to me that it is no longer needed. But I'm not 100% sure and want to know more details here.

Why does es usually need index aliases? 
As a rule, in order to conveniently group several indexes and have several indexes, access all of them at once through a common alias. 
As I see, this is not our case. The old course-discovery has one index from the point of view of the application.

**My understanding of the need for aliases in course-discovery is:**
They are needed to be able to drop old indexes at runtime and create new ones on the fly. A timestamp is added to the index names. This is done so that you can delete old indexes by specifying the time, i.e. remove all indexes that were created before the given label. And the application itself refers to the alias and does not notice this variety of index names and always knows which index to work with. This is done as I understand it to manage disk space efficiently.
This is my short understanding of what aliases are for here.
Is it so?
If so, then it seems to me it's overhead. After all, it is much easier to reset (delete) the index and start repopulating data from the sql database. This is handy for Django-es-dsl. This may require downtime, which can be critical. 
Also, the data in the index is always up to date since Django-es-dsl listens for django signals.
But I assume that I may not know anything.

Now as I wrote, this is a blocker for me and I need to understand how to do it. I will ask about this at our meeting, so if you answer here, we will save time.

Reply:
I believe that the alias is mostly used to control how/when we update the index.

In `update_index.py`, we first build a new index (but don't use it yet, by not yet pointing the alias at it). If the new index is not acceptable (there is some sanity check code to make sure we didn't accidentally drop all the items in the index), we never update the alias.

So I think the alias is really just about that -- making sure we don't immediately use a new index in case it's bad, and then switching at once to the new index if it passes muster.

DSL's approach is signal based and always up to date? That sounds like a very different model indeed. Is there any room there for a similar sort of "sanity check?"

Reply:
> DSL's approach is signal based and always up to date?

Here, yes and no. I see that the signal-based approach cannot be used in the course-discovery service itself. Here I found such a message https://github.com/raccoongang/course-discovery/blob/master/course_discovery/settings/base.py#L528-L530. haystack can also use signals, it has everything for this, but as we can see from the message, it was decided to refuse signals. So in django-es-dsl we will also use `BaseSignalProcessor`, i.e. not signals.

Now about `"sanity check"`....

I think I understand what you mean. I sat and thought well to understand why this case appeared at all, why we are so afraid to create the wrong index, with incorrect data. And developers came up with such a safety net. To be honest, I really like the implementation of this net and I respect the author who invented it 🙂

So this is why it is needed ... I think that in general, this problem arose due to the complexity of the `update_index` command itself https://django-haystack.readthedocs.io/en/latest/management_commands.html#update-index . There are a lot settings in it, and it is true that it is possible to accidentally create incomplete, wrong data. It can get confusing. Also, in each index there is such a method for haystack `get_updated_field`, if you make a mistake somewhere in it, you can also create the wrong data. And that could be a problem. Plus, as I see, before the index was common (that is, it is one) for all four models `Course`,` CourseRun`, `Program`,` Person`. Those. losing all data from one index at once and then restoring them again is an expensive and heavy operation, and here we need to be sure and do not do stupid things. I understand.

But....

**First**, we switched to django-elasticsearch-dsl and here this library offers the following commands https://django-elasticsearch-dsl.readthedocs.io/en/latest/management.html. Here's a look at implementation https://github.com/django-es/django-elasticsearch-dsl/blob/master/django_elasticsearch_dsl/management/commands/search_index.py It’s great and beautiful. I have investigated the source code carefully and everything here is very reliable, smart, and simple. There is no place to go wrong. 
It is immediately clear that these commands are very simple and different, i.e. you can choose exactly the one you need. You can just repopulate the index from the SQL database, we can do a rebuild, i.e. delete the old index, create a new one, and put all the necessary data from the SQL database there. Please note that only one request to elasticsearch is used to put data because we are dealing with bulk request https://www.elastic.co/guide/en/elasticsearch/reference/master/docs-bulk.html, i.e. prepare general json and put everything at once in elasticsearch database. Plus, the data can be split into chunks, and requests to es can be executed in parallel, which does not overload the network with a fat HTTP request. Those. We can work flexibly and quickly.

**Second**, we do not have a common index now. We, as elasticsearch recommends in its best practices, have four indices(each model separately), which means that we can create, delete, populate each index separately at any time, which is not so expensive in terms of performance problems.

**Third**, es's philosophy is that we shouldn't be afraid of losing data in its database. After all, this data can be quickly and conveniently restored from sql database, i.e. persistent, master database. After all, es database is needed not for storing data, but for fast search, full-text search.

So what do I want to say? I don't see the need for a "sanity check" at all when we use django-es-dsl and in our conditions. I really thought about how it could help, but I couldn't find now a real reason to do so. I think that it is no longer necessary and it will be overengineering.
Yes, I can implement for Django-es-dsl, i.e. to do the same "sanity check", but it will be just a waste of time in my opinion.
What do you think Michael?

Reply:
I can see both sides. I think on one hand, the sanity check is very akin to a unit test. It can feel overly cautious in the moment sometimes, but it helps give confidence when making future changes. In this case, as either DSL or course-discovery changes around this code, we have basically a runtime unit test checking that our update_index code still works. Because if we don't notice that something went wrong, the only way we'd hear about it is support tickets for people not finding anything when searching.

Also, in some contexts (non-edx.org installations of open edx, mostly), course-discovery is less of a source-of-truth for course data and Studio is instead. In those cases, when course-discovery runs `refresh_course_metadata`, it pulls in all the courses from Studio and makes sure the databases line up. If that process goes wrong, this sanity check would maybe catch a problem like that (though, in that case... the index is now out of date and we have bigger problems with the db and that seems like a very indirect way to catch the problem, but I think that was maybe part of the original idea).

That said, it is added complexity that we probably don't need, and we would likely be able to recover fast from any temporary failure anyway, as you say.

What about a half-way measure? If the whole alias switching is harder to support in DSL land, could we at least check the index size before and after an update, and if the change is too big, raise a CommandError in the `update_index` job with a message like "Index changed more than we expected, we are failing the job so that you will notice and investigate in case this is a sign of a problem."

That way, we will get alerted if something (not sure what) goes wrong, sooner than support tickets coming in would do. And hopefully that would be a pretty easy check to add.

Reply:
Michael, thanks for suggesting the half-way measure. I will keep this option in my head. 
But I see that we have written so much text and words that this feature deserves to be implemented. I think we can try to do it. If I see that it is long and difficult, then already `half-way measure`.